File: src/main/kotlin/pages/TagsInputBoxPage.kt
package pages

import com.microsoft.playwright.Locator
import helpers.PlaywrightObject.page
import io.qameta.allure.Step

class TagsInputBoxPage {
    private val inpText get() = page?.locator("input[type='text']")
    private val listTags get() = page?.locator("//li")?.all()
    private val txtCount get() = page?.locator("//div[@class='details']/p/span")
    private val btnRemoveAll get() = page?.locator("button")

    fun open(): TagsInputBoxPage {
        page?.navigate("${System.getProperty("TEST_HOST")}/apps/tags-input-box/")
        return this
    }

    @Step("Add Tag")
    fun addTag(tag: String): TagsInputBoxPage {
        inpText?.fill(tag)
        inpText?.press("Enter")
        return this
    }

    @Step("Get all Tags")
    fun getTags(): List<TagObject> {
        return listTags!!.map { Tag(it) }
            .map {
                TagObject(
                    text = it.name,
                    button = it.button
                )
            }
    }

    @Step("Get Tags Count")
    fun getTagsCount(): Int {
        return txtCount!!.innerText().toInt()
    }

    @Step("Remove All Tags")
    fun removeAllTags(): TagsInputBoxPage {
        btnRemoveAll!!.click()
        return this
    }
}

class Tag(item: Locator) {
    val name = item.innerText()
    val button = item.locator("xpath=.//i")
}

data class TagObject(
    val text: String,
    val button: Locator,
)

File: pages/TagsInputBoxPage.ts
import { BasePage } from '../helpers/BasePage'

export class TagsInputBoxPage extends BasePage {
    inpText = this.page.locator("input[type='text']")
    listTags = this.page.locator("//li")
    txtCount = this.page.locator("//div[@class='details']/p/span")
    btnRemoveAll = this.page.locator("button")

    async open(): Promise<TagsInputBoxPage> {
        await this.page.goto("https://qaplayground.dev/apps/tags-input-box/")
        return this
    }

    async addTag(tag: string): Promise<TagsInputBoxPage> {
        await this.inpText.fill(tag)
        await this.inpText.press("Enter")
        return this
    }

    async getTagsCount(): Promise<number> {
        return await parseInt(await this.txtCount.innerText(), 10)
    }

    async removeAllTags(): Promise<TagsInputBoxPage> {
        await this.btnRemoveAll.click()
        return this
    }

    async getTags(): Promise<{ text: string, button: import("@playwright/test").Locator }[]> {
        return await Promise.all((await this.listTags.all()).map(async (tag) => {
            return { text: await tag.innerText(), button: await tag.locator(".//i") }
        }))
    }
    
}   
